src/MOL_discretization.jl

```diff
@@ -1,22 +1,220 @@
+    if input isa ModelingToolkit.Constant
+        return :($input.value)
+    elseif input isa Operation
+        if input.op isa Variable
+            if haskey(lhs_nonderiv_depvars,input.op)
+                x = lhs_nonderiv_depvars[input.op]
+                if x isa ModelingToolkit.Constant
+                    expr = :($x.value)
+                else
+                    expr = Expr(x)
+                    expr = :(x=i*$dx;eval($expr))
+                end
+            elseif grade == 1
+                # TODO: the discretization order should not be the same for
+                #       first derivatives and second derivarives
+                j = findfirst(x->x==input.op, lhs_deriv_depvars)
+                expr = :((u[$j][i]-u[$j][i-1])/$dx)
+            elseif grade == 2
+                j = findfirst(x->x==input.op, lhs_deriv_depvars)
+                expr = :((u[$j][i+1]-2.0*u[$j][i]+u[$j][i-1])/($dx*$dx))
+            else
+                j = findfirst(x->x==input.op, lhs_deriv_depvars)
+                expr = :(u[$j][i])
+            end
+            return expr
+        elseif input.op isa Differential
+            grade += 1
+            return discretize_2(input.args[1],grade,order,dx,lhs_deriv_depvars,lhs_nonderiv_depvars)
+        elseif input.op isa typeof(*)
+            expr1 = discretize_2(input.args[1],grade,order,dx,lhs_deriv_depvars,lhs_nonderiv_depvars)
+            expr2 = discretize_2(input.args[2],grade,order,dx,lhs_deriv_depvars,lhs_nonderiv_depvars)
+            return Expr(:call,:*,expr1,expr2)
+        elseif input.op isa typeof(/)
+            expr1 = discretize_2(input.args[1],grade,order,dx,lhs_deriv_depvars,lhs_nonderiv_depvars)
+            expr2 = discretize_2(input.args[2],grade,order,dx,lhs_deriv_depvars,lhs_nonderiv_depvars)
+            return Expr(:call,:/,expr1,expr2)
+        elseif input.op isa typeof(-)
+            if size(input.args,1) == 2
+                expr1 = discretize_2(input.args[1],grade,order,dx,lhs_deriv_depvars,lhs_nonderiv_depvars)
+                expr2 = discretize_2(input.args[2],grade,order,dx,lhs_deriv_depvars,lhs_nonderiv_depvars)
+                return Expr(:call,:-,expr1,expr2)
+            else #if size(input.args,1) == 1
+                expr1 = discretize_2(input.args[1],grade,order,dx,lhs_deriv_depvars,lhs_nonderiv_depvars)
+                return Expr(:call,:*,:(-1),expr1)
+            end
+        elseif input.op isa typeof(+)
+            if size(input.args,1) == 2
+                expr1 = discretize_2(input.args[1],grade,order,dx,lhs_deriv_depvars,lhs_nonderiv_depvars)
+                expr2 = discretize_2(input.args[2],grade,order,dx,lhs_deriv_depvars,lhs_nonderiv_depvars)
+                return Expr(:call,:+,expr1,expr2)
+            else #if size(input.args,1) == 1
+                expr1 = discretize_2(input.args[1],grade,order,dx,lhs_deriv_depvars,lhs_nonderiv_depvars)
+                return Expr(expr1)
+            end
+        end
+    end
+end

```

Comment:
All of this code is quite unnecessary. It should just be a call to `CenteredDifference` and `UpwindDifference`

Reply:
The aim of this algorithm is to traduce an rhs such as: 
```
ln( v(t,x)^Dx(phi(t,x)) / (u(t,x)*(1-u(t,x))) )
where, v(t,x) = 1+2*t*x
```
to an expression like this:
```
ln(  (1+2*t*i*0.025)^((u[1][i]-u[1][i-1])/0.025) / (u[1][i]*(1-u[1][i])) )
```
where x=i * dx=i * 0.025, and "u" stores each variable (phi has index 1). 

The obtained expression is then evaluated at each time step in the function "f(du,u,p,t)". Using this strategy I avoid using matrices, which is convenient in non-linear problems, and I store every variable within "u", which is consistent with the "f" interface. How would you translate the rhs that I gave using CenteredDifference?


On the other hand, last month I have been thinking how to reimplement this prototype code in a simpler way. Here is an alternative:

Assuming lhs has expressions of this form: Dx(phi(t,x)) or v(t,x)

For each rhs (e.g. Dx(v(t,x)*phi(t,x))) associated to an lhs variable (e.g. phi)
1. expand derivatives (does mtk provide this feature?)
```
     e.g.: Dx(v(t,x)*phi(t,x))
             => Dx(v(t,x))*phi(t,x)+v(t,x)*Dx(phi(t,x))
```
2. replace derivatives and variables with its discretizations
```
     e.g.: Dx(v(t,x))*phi(t,x)+v(t,x)*Dx(phi(t,x))
            => ((u[1][i]-u[1][i-1])/0.025)*u[2][i]+u[1][i]*((u[2][i]-u[2][i-1])/0.025)
```
3. simplify obtained expression (does mtk provide this feature?)
4. solve the ODE problem

Reply:
>to an expression like this:

You don't want the expression though, since with that form you cannot hit the stencil compiler. This is the reason for the operator form since otherwise you will not have speed or GPU-compatibility.

>How would you translate the rhs that I gave using CenteredDifference?

That's an upwind operator call and then a broadcast expression.

Reply:
> You don't want the expression though, since with that form you cannot hit the stencil compiler. This is the reason for the operator form since otherwise you will not have speed or GPU-compatibility.

The whole code is based on building the discretizations as expressions that can be evaluated within f(du,u,p,t), avoiding expressions implies reimplementing many things.

> That's an upwind operator call and then a broadcast expression.

Could you please explicitly write the outcome you expect to get when an Operation like the one below is discretized?:
```
ln( v(t,x)^Dx(phi(t,x)) / (u(t,x)*(1-u(t,x))) )
```

Reply:
`Dx(phi(t,x))` would be a forward difference operator, and the rest would be a broadcast expression

Reply:
> Dx(phi(t,x)) would be a forward difference operator, 

That I had understood.

> and the rest would be a broadcast expression

Here is where I do not follow you, would you please write a single line of code with the discretized rhs, using a broadcast expression and the UpwindDifference operator?

```
rhs = ln( v(t,x)^Dx(phi(t,x)) / (u(t,x)*(1-u(t,x))) )
discretized_rhs = ...
```

This code should discretize the derivative, with UpwindDifference, but also the other variables: v and u. Furthermore, this discretized code (discretized_rhs) should be evaluated w.r.t. t at each time iteration, that is one of the reasons I used expressions to build the discretization.